README.md
# IWAction Versão 1.0.05  12/05/2020
**Uma biblioteca para ser usada com Scriptcase.**

Video explicativo [youtube.com](https://youtu.be/3aI-8MgtbYE)

Conheça [Scriptcase](https://scriptcase.com.br)

Discussão em [Fórum](https://forum.scriptcase.com.br/)

Grupo [Telegram](https://t.me/joinchat/F1fMxkV3oNu6JVFxfJ1Xqg)

Grupo [WhatsApp](https://chat.whatsapp.com/I3sg8dHFfQqKJYjVgtQkM8)


# O que é IWAction Classe PHP?

São um conjuntos de scripts php, js, html e css para você poder criar uma barra de ferramentas em linha nas aplicações do tipo Grid no Scripcase.

O que faz essa Biblioteca?

* Adiciona texto, imagem (da Galeria de Imagens do Scriptcase ou de sua própria Galeria),  FontAwaesome 
* Permite link com aplicação Scriptcase ou uma URL abrindo  aplicação sobre a aplicação Grid corrente.
* Permite Link com aplicação Scriptcase ou URL abrindo em modal



# Instalação


**Na aplicação Grid** 

evento **onscriptinit**

```php
sc_include_library('prj','grid','actions-master/IWActions.class.php');
sc_include_library('prj','grid','actions-master/include.php');
```

evento **onrecord**

```php

$line = {sc_seq_register};

// Instanciando a Classe
$tb = new IWActions($line);

// Condição para apresentação da ação/imagem ou não
$tb->setCondition({campo}=='S'); 

// Gera Espaço entre os itens da barra de ferramentas
$tb->setItemSpace(5); 

// Utiliza uma barra separadora entre os itens (dark, light)
// Parâmtro 1 valores dark ou light (padrão)
$tb->setSeparator('dark'); 

// Acrescenta uma borda pontilhada abaixo do texto
// Sem parâmetros
$tb->setBorderBottom();

// Apresenta um texto no item . Parâmetros Posicionais
// Parâmetro 1 para condição verdadeira, 
// Parâmetro 2 para falsa (Opcional).
// Use setColor para atribuir cor as fontes
// Uset setTextStyle para atribuir etilos aos textos
$tb->setText('Linha 1', 'Linha ~'); 

// Atribui cores ao texto ou a Fontes Awesome
// Parâmetro 1 para item apresentado na condição verdadeira, 
// Parâmetro 2 para item apresentado na condição falsa (Opcional).
$tb->setColor('blue','red');

// Atribui estilos ao texto (qualquer estilo CSS aplicável a texto)
// Parâmetro 1 para item apresentado na condição verdadeira, 
// Parâmetro 2 para item apresentado na condição falsa (Opcional).
$tb->setTextStyle('color=blue, font-size=24', 'color=red, font-size=20');

// Muda o estilo do cursor ao item, veja estilos em [W3C](https://www.w3schools.com/cssref/pr_class_cursor.asp)
// Parâmetro 1 para item apresentado na condição verdadeira, 
// Parâmetro 2 para item apresentado na condição falsa (Opcional).
$tb->setCursor('pointer', 'help');

// Apreenta ToolTip acima do item
// Parâmetro 1 para item apresentado na condição verdadeira, 
// Parâmetro 2 para item apresentado na condição falsa (Opcional).
$tb->setToolTip('Texto','Texto');

// Apreenta ToolTip definido pelo método setToolTip a esquerda do item 
// Não utiliza parâmetros
$tb->setToolTipLeft();

// Apreenta ToolTip definido pelo método setToolTip abaixo do item 
// Não utiliza parâmetros
$tb->setToolTipBottom();

// Fecha o item e prepara para novo item na barra
$tb->close(); 

// Determina altura da Imagem
$tb->setImageHeight(24); 

// busca imagem na Galeria de imagens do Scriptcase 
// Parâmetros Posicionais
// Parâmetro 1: Imagem para Condição Verdadeira
// Parâmetro 2: (opcional) (Padrão: grp,img) categoria e tipo da imagem na galeria de Imagens do Scriptcase
// Parâmetro 3: (Opcional) Imagem da Galeria para Condição Falsa 
// Parâmetro 4: (opcional) (Padrão: mesmo do Parâmetro 2) categoria e tipo da imagem na galeria de Imagens do Scriptcase
$tb->setSCImage('carpeta_edita.png','sys,ico','cadeado.jpg');

$tb->close();

// MOnta link para acesso a aplicação baseado no Scriptcase 
$link = sc_make_link(aplicação,par1=val1;paar2=val2);

// Informa que item ser Link de chamada a aplicação, pode ser URL
// Abre no destino atual (_self) **Previsão de melhorias**
// Parâmetros Posicionais
// Parâmetro 1: Link para Condição Verdadeira
// Parâmetro 2: (Opcional) Link para Condição Falsa
$tb->setLink($link);

// Abre Link em Modal e atribue True a propriedade modal [$tb->Modal(true);]
// Parâmetros Nomeados informados por string todos opcionais
// Padrão se não Informado: 'maxWidth=80%, width=80%,background=#000,padding=5px,textAlign=center,height=80%'
// maxWidth (max-width) : Largura máxima da Modal
// width : Largura da Modal
// height : Altura da Modal
// background : Côr em Hexa ou Nominal do fundo da Modal
// padding : Espaçamento entre as margens e o conteúdo da Modal
// textAlign (text-align) : Alinhando do conteúdo da Modal
$tb->setModalStyle('width=40%');

// Informa imagem da Galeria pessoal do desenvolvedor adicionadas a pasta img no raiza da biblioteca externa criada
// Parâmtros posicionais
// Parâmtro 1 : Imagem para Condição Verdadeira
// Parâmtro 2 : (Opcional) Imagem para condição Falsa
$tb->setImage('editar.png');

$tb->close();

// Informa que o item da Barra de Ferramentas terá como exibição uma Fonte Awesome
// Parametros Posicionais
// Parâmetro 1 : Classe de Estilo para fonte em caso de COndição Verdadeira
// Parâmetro 2 :  (Opcional) Fonte Awesome para Concidção Falsa
// User setColor para atribuir cor as fontes
$tb->setAwesome('far fa-thumbs-up','far fa-thumbs-down');

$tb->close();

// Retorno Html da barra de ferramentas
// Fecha a Barra de Ferramentas e Deixa Preparado para crianão de outra Barra.
{action} = $tb->createToolBar();
ou
{action} = $tb->create();
```

# Notas:

**Parâmtros Posicionais** Cada parâmetro tem sua posição determinada conforme seu conteúdo

**Parâmtros Nomeados** O PHP não suporta (ainda parâmtros nomeados), então criei um Método Mágico para possibilitar informar os parâmtros pelos seus nomes sem a obriagação de posição exata (Na Classe o Método se chama `parsToObj($pars, $default)`)


**IMPLEMENTAÇÕES FUTURAS**
* Apresentação de Tooltip ao passar o mouse sobre os itens
* Ação do evento ajax onclick da Grid
* Dialog de Confirmação em SweetAlert antes da execução da Ação em evento Ajax

**Métodos e Bibliotecas de outros autores utilizados**

`get_leaf_dirs($dir)` por [Anton Backer 2006](https://www.php.net/manual/pt_BR/function.dir.php#60374)
Renomeada para `recursiveDir($dir)`

`jquery-modal` por [Kyle Fox](https://github.com/kylefox/jquery-modal)

**NÃO PIRATEIE CÓDIGOS DE OUTROS, RESPEITE OS DIREITOS AUTORAIS**


CHANGELOG.md
# CHANGELOG
**Release 1.0.07 20/05/2020**

    * NOVO - Implementado Método setBorderBottom
    * NOVO - Implementado Método setToolTipWordWrap
    * MELHORIA - Implementado Método create com abreviação de createToolBar
    * NOVO - Implementado Loading - Carregando

**Re1ease 1.0.06 13/05/2020** 

    * BUG - Corrigido nome do método setToolTipBottom, o qual estava se referenciando incorretamente a
    classe de estilos para apresentar o ToolTip abaixo do item.


**Re1ease 1.0.05 12/05/2020 Requer atualização completa** 

    * NOVO - Implementado Novo Método: setCursor 
    * NOVO - Implementado Novo Método: SetColor
    * NOVO - Implementado Novo Método: SetTextStyle
    * NOVO - Implementado Novo Método: setToolTip, setTollTipBottom, setToolTipLeft
    * AJUSTE - Alterado Método setAwesome: Não existem mais os parâmetros de cores, ver [README.md]()
    * MELHORIA no método interno parsToObj: Aceita um '-' no nome do parâmetro.
    * OTIMIZAÇÃO DE RECURSOS DE PROCESSAMENTO E MEMÓRIA - Todos os Métodos sofreram otimização

**Re1ease 1.0.03 09/05/2020** 

    * Correções para otimização

**Re1ease 1.0.00  08/05/2020**

    * Liberação da Biblioteca


include.php
<?php
echo <<<HTML
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-modal/0.9.1/jquery.modal.min.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/jquery-modal/0.9.1/jquery.modal.min.css" />
<style type="text/css">
a.action-link {
  text-decoration: none;
}

a.action-link:hover {
  text-decoration: none;
  	border-bottom: 1px dotted #7DB6E2;
}

.modal a.close-modal[class*="icon-"] {
  top: -10px;
  right: -10px;
  width: 20px;
  height: 20px;
  color: #fff;
  line-height: 1.25;
  text-align: center;
  text-decoration: none;
  text-indent: 0;
  background: #900;
  border: 2px solid #fff;
  -webkit-border-radius: 26px;
  -moz-border-radius: 26px;
  -o-border-radius: 26px;
  -ms-border-radius: 26px;
  -moz-box-shadow:    1px 1px 5px rgba(0,0,0,0.5);
  -webkit-box-shadow: 1px 1px 5px rgba(0,0,0,0.5);
  box-shadow:         1px 1px 5px rgba(0,0,0,0.5);
}

.tooltip {
	position: relative;
	display: inline-block;
}

.borderBottom {
	border-bottom: 1px dotted black;
}

.tooltip .tooltiptext {
  visibility: hidden;
  width: auto;
  background-color: black;
  color: #fff;
  text-align: center;
  border-radius: 6px;
  padding: 5px;
  position: absolute;
  z-index: 1;
  bottom: 150%;
  left: 50%;
  margin-left: -60px;
	opacity: 0;
	transition: opacity 0.7s;
	word-wrap: break-word;
}

.tooltip .tooltiptext::after {
  content: "";
  position: absolute;
  top: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-color: black transparent transparent transparent;
}

.tooltip:hover .tooltiptext {
  visibility: visible;
	opacity: 1;
}


.tooltip .tooltiptext_left {
  visibility: hidden;
  width: auto;
  background-color: black;
  color: #fff;
  text-align: center;
  border-radius: 6px;
  padding: 4px;
  position: absolute;
  z-index: 1;
  top: 0px;
  right: 110%;

}

.tooltip .tooltiptext_left::after {
  content: "";
  position: absolute;
  top: 50%;
  left: 100%;
  margin-top: -5px;
  border-width: 4px;
  border-style: solid;
  border-color: transparent transparent transparent black;
}

.tooltip:hover .tooltiptext_left {
  visibility: visible;
}

.tooltip .tooltiptext_bottom {
  visibility: hidden;
  width: auto;
  background-color: black;
  color: #fff;
  text-align: center;
  border-radius: 6px;
  padding: 5px;
  position: absolute;
  z-index: 1;
  top: 150%;
  left: 50%;
  margin-left: -60px;
}

.tooltip .tooltiptext_bottom::after {
  content: "";
  position: absolute;
  bottom: 100%;
  left: 50%;
  margin-left: -5px;
  border-width: 5px;
  border-style: solid;
  border-color: transparent transparent black transparent;
}

.tooltip:hover .tooltiptext_bottom {
  visibility: visible;
}
</style>
<div  id="modal-form"  class = "modal">
  <iframe id=modal-form-iframe  style='allowfullscreen:true;border:0;width:100%'></iframe>
</div>
<script>
$(document).ready(function () {
     $('a[href="#login-form"]').click(function(event) {
        event.preventDefault();
        $(this).modal({
            escapeClose: false,
          clickClose: false,
          showClose: true,
          fadeDelay: 0.50
        });
        $('#modal-form-iframe').css('height',$(".modal").height()-40);
    });
});
function modalHtml(html) {
   $('#modal-form').html(html);
}
function modalIframeSrc(src) {
    $('#modal-form-iframe').attr('src', src);
    $('#modal-form').modal();
}

function modalStyle(style) {
    $('#modal-form').attr('style', style);
    $('#modal-form-iframe').css('height', $(".modal").height() - 40);
}
</script>
HTML;


IWActions.class.php
<?php
/* Classe iwActions
	Classe para criação de botões de ação nas linhas de uma Grid Scriptcase
	Autor: Haroldo B Passos / InfinitusWeb
	Copyrirght: 2020 / 2020 Versão: 1.00.004 
	Fase II : Versão Beta Teste 13/05/2020 1.01.006 ...
*/

class IWActions
{
	private $lineSeq;
	private $item = '';
	private $id = 0;
	private $condition = TRUE;
	private $imageHeight = 20;
	private $pathImage;
	private $pathAction;
	private $itemSpace = '2';
	private $color = '';
	private $textStyle = '';
	private $text = false;
	private $toolBar = [];
	private $borderBottom = false;
	private $toolTip = false;
	private $itemSeparator = "<div Style ='float: left; display: inline-block;padding-right: 2px'></div>";
	private $link = '';
	private $modal = false;
	private $modalStyle = 'max-width: 80%; width: 40%; background: rgb(0, 0, 0); padding: 5px; text-align: center; height: 80%';
	private $cursor = '';
	private $toolTipWordWrap =  false;

	function __construct($lineSeq = 0)
	{
		$this->lineSeq = $lineSeq;
		$dirs = $this->recursiveDir('../_lib/libraries');
		foreach ($dirs as $path) {
			$pos = strpos($path, 'actions-master');
			if (is_numeric($pos)) {
				break;
			}
		}
		$path = substr($path, 0,  $pos - 1);
		$this->pathImage = $path . '/img';
		$this->pathAction = $path . '/actions-master';
	}

	public function setItemSpace($itemSpace)
	{
		$this->itemSpace = $itemSpace;
		$this->itemSeparator = "<div Style ='float: left; display: inline-block;padding-right:{$this->itemSpace}px'></div>";
	}

	public function setSeparator($flag = 'light')
	{
		if ($flag) {
			$space = $this->itemSpace / 2;
			$image = $flag === 'dark' ? 'separator-dark.png' : 'separator-light.png';
			$this->itemSeparator = "<div Style = 'float: left; display: inline-block;'><img src='{$this->pathAction}/img/$image' height='{$this->imageHeight}px' style='padding: 0px {$space}px;' /></div>";
		} else {
			$this->itemSeparator = "<div Style ='float: left; display: inline-block;padding-right: {$this->itemSpace}px'></div>";
		}
	}

	public function setBorderBottom()
	{
		$this->borderBottom = true;
	}

	public function setImageHeight($imageHeight)
	{
		$this->imageHeight = $imageHeight;
	}

	public function setCondition($condition)
	{
		$this->condition = $condition;
	}

	public function setColor($colorTrue, $colorFalse = '')
	{
		$this->color = ($this->condition) ? $colorTrue : $colorFalse;
	}

	public function setSCImage($imgTrue, $categTrue = 'grp,img', $imgFalse = '', $categFalse = '')
	{
		if ($this->condition) {
			$categ = explode(',', $categTrue);
			$image = $imgTrue;
		} else {
			$categ = empty($categFalse) ? explode(',', $categTrue) : explode(',', $categFalse);
			$image = $imgFalse;
		}
		if ($image) {
			$categ[0] =  strtolower($categ[0]) == 'prj' ? 'grp' : strtolower($categ[0]);
			$image = $categ[0] . '__NM__' . strtolower($categ[1]) . '__NM__' . $image;
			$this->item = "<img src='../_lib/img/$image' height='{$this->imageHeight}px'/>";
		}
	}

	public function setImage($imageTrue, $imageFalse = '')
	{
		$path = $this->pathImage . '/';
		$image = ($this->condition) ? $imageTrue : $imageFalse;
		if ($image) {
			$this->item = "<img src='$path{$image}'  height='{$this->imageHeight}px' />";
		}
	}

	public function setAwesome($fontAwesomeTrue, $fontAwesomeFalse = '')
	{
		$fontAwesome = ($this->condition) ? $fontAwesomeTrue : $fontAwesomeFalse;
		$color = ($this->color) ? 'color: ' . $this->color : '';
		$this->item  = "<span style='font-size:{$this->imageHeight}px;$color;'><i class='$fontAwesome'></i></span>";
	}

	public function setTextStyle($styleTrue, $styleFalse = '')
	{
		$this->textStyle = ($this->condition) ? $this->parsToStyle($styleTrue) : $this->parsToStyle($styleFalse);
	}

	public function setText($textTrue, $textFalse = '')
	{
		$this->text = true;
		$this->item = ($this->condition) ? $textTrue : $textFalse;
	}

	public function setLink($linkTrue, $linkFalse = '')
	{
		$this->link = ($this->condition) ? $linkTrue : $linkFalse;
	}

	public function setModal($flag = true)
	{
		$this->modal = $flag;
	}

	public function setModalStyle($value = '')
	{
		$p = $this->parsToObj($value, 'maxWidth=80%, width=40%,background=#000,padding=5px,textAlign=center,height=80%');
		$this->modalStyle = "max-width: {$p->maxWidth}; width:{$p->width};height:{$p->height};background:{$p->background};padding:{$p->padding};text-align:{$p->textAlign}";
		$this->modal = true;
	}

	public function setCursor($cursorTrue = 'default', $cursorFalse = '')
	{
		$cursor = $this->condition ? $cursorTrue : $cursorFalse;
		$cursor = ($cursor) ? 'cursor:' . $cursor : '';
		if ($cursor) $this->cursor = $cursor;
	}

	public function setToolTipBottom()
	{
		$this->toolTip[1] = 'tooltiptext_bottom';
	}
	public function setToolTipLeft()
	{
		$this->toolTip[1] = 'tooltiptext_left';
	}

	public function setToolTipWordWrap($nChars = false)
	{
		$this->toolTipWordWrap = $nChars;
	}

	public function setToolTip($hintTrue, $hintFalse = '')
	{
		$this->toolTip[1] = (isset($this->toolTip[1])) ? $this->toolTip[1] : 'tooltiptext';
		$hint = ($this->condition) ? $hintTrue : $hintFalse;
		$this->toolTip[0] = $hint;
	}

	public function close()
	{
		$borderBottom = ($this->borderBottom) ? ' borderBottom' : '';

		if ($this->text) {
			$style = $this->textStyle;
			$color = ($this->color) && strpos($style, 'color') === false ? "color: {$this->color};" : '';
			$style = ($color || $style) ? "$style{$color}" : '';
			$class = ($this->borderBottom) ? "class=$borderBottom" : '';
			$this->item =  "<span $class $style>{$this->item}</span>";
			$borderBottom = $borderBottom ?: '';
		}

		if ($this->link) {
			if ($this->modal) {
				$a = "href='#modal-form' onmouseup=\"modalStyle('{$this->modalStyle}');modalIframeSrc('{$this->link}');\"  "; //rel=modal:open
			} else {
				$a = "href='{$this->link}'";
			}
			$this->item = "<a class='action-link{$borderBottom}' $a>{$this->item}</a>";
			$borderBottom = $borderBottom ?: '';
		}

		if ($this->toolTip[0]) {
			$style = "Style = 'float: left; {$this->cursor}'";
			$this->toolTip[0] = ($this->toolTipWordWrap) ? wordwrap($this->toolTip[0], $this->toolTipWordWrap, "<br/>\n") : $this->toolTip[0];
			$this->item = "<div class='tooltip' $style>{$this->item}<span class='{$this->toolTip[1]}'>{$this->toolTip[0]}</span></div>";
		} else {
			$style = "Style = 'float: left; display: inline-block;{$this->cursor}'";
			$class = ($this->borderBottom) ? "class=$borderBottom" : '';
			$this->item = "<div $class $style>{$this->item}</div>";
		}

		$this->toolBar[$this->id] = $this->item;
		$this->id++;
		$this->setReset();
	}

	private function setReset($total = false)
	{
		$this->link = '';
		$this->cursor = '';
		$this->color = '';
		$this->textStyle = '';
		$this->text = false;
		$this->toolTip = false;
		$this->borderBottom = false;
		$this->modal = false;
		$this->modalStyle = 'max-width: 80%; width: 40%; background: rgb(0, 0, 0); padding: 5px; text-align: center; height: 80%';
		$this->item = '';
		$this->toolTipWordWrap =  false;

		if ($total) {
			$this->id = 0;
			$this->condition = TRUE;
			$this->imageHeight = 20;
			$this->itemSpace = '2';
			$this->toolBar = [];
			$this->separator = false;
			$this->imgSeparator = '';
		}
	}

	public function createToolBar()
	{
		$html = '';
		foreach ($this->toolBar as $key => $value) {
			$strip_tags = trim(strip_tags($value, '<img><i>'));
			$value = ($strip_tags) ? $value : '';
			$html .= ($key > 0 && !empty($value)) ? $this->itemSeparator . $value : $value;
		}
		$this->setReset(true);
		return $html;
	}

	public function create()
	{
		return $this->createToolBar();
	}

	public function getPathAction()
	{
		return $this->pathAction;
	}

	//By Anton Backer 2006 get_leaf_dirs($dir) 
	private function recursiveDir($dir)
	{
		$array = array();
		$d = dir($dir);
		while (false !== ($entry = $d->read())) {
			if ($entry != '.' && $entry != '..') {
				$entry = $dir . '/' . $entry;
				if (is_dir($entry)) {
					$subdirs = $this->recursiveDir($entry);
					if ($subdirs)
						$array = array_merge($array, $subdirs);
					else
						$array[] = $entry;
				}
			}
		}
		$d->close();
		return $array;
	}

	private function parsToStyle($style)
	{
		$style = "style='" . str_replace(['=', ','], [':', ';'], $style) . "'";
		return $style;
	}

	//Método Mágico -> permite utilizar parametros nomeados nos demais métodos da classe
	//By Haroldo
	private function parsToObj($pars, $default)
	{
		$pars = !empty($pars) ? explode(',', $pars) : [];
		$res = [];
		foreach ($pars as $value) {
			$arr = explode('=', $value);
			if (is_numeric(strpos($arr[0], '-'))) {
				$atr = explode('-', $arr[0]);
				$arr[0] = $atr[0] . ucfirst($atr[1]);
			}
			$arr[0] = trim($arr[0]);
			$arr[1] = trim($arr[1]);
			if (strpos($arr[1], ';')) {
				$arr[1] = explode(';', $arr[1]);
			}
			$res[$arr[0]] = $arr[1];
		}
		$pars = $res;
		$res = [];
		$default = explode(',', $default);
		foreach ($default as $value) {
			$arr = explode('=', $value);
			$arr[0] = trim($arr[0]);
			$arr[1] = trim($arr[1]);
			if (strpos($arr[1], ';')) {
				$arr[1] = explode(';', $arr[1]);
			}
			$res[$arr[0]] = $arr[1];
		}
		$json = json_encode(array_merge($res, $pars));
		return json_decode($json);
	}
}
